File: solutions/typescript/diamond/1/diamond.ts
export function makeDiamond(character: string): string {
  const buildShape = (letter: string): string => {
    const width = (letter.charCodeAt(0) - 'A'.charCodeAt(0)) * 2 + 1;

    const buildLine = (letter: string, width: number): string => {
      const innerWidth = (letter.charCodeAt(0) - 'A'.charCodeAt(0)) * 2 - 1;
      const outerWidth = Math.floor((width - ((letter.charCodeAt(0) - 'A'.charCodeAt(0)) * 2 + 1)) / 2);
      const letterStr = letter;
      const outerPad = ' '.repeat(outerWidth);

      const content = innerWidth < 1
        ? letterStr
        : letterStr + ' '.repeat(innerWidth) + letterStr;

      return outerPad + content + outerPad;
    };

    const front = Array.from({ length: letter.charCodeAt(0) - 'A'.charCodeAt(0) + 1 }, (_, i) => 
      buildLine(String.fromCharCode('A'.charCodeAt(0) + i), width)
    );

    return front.concat(front.slice(0, -1).reverse()).join('\n') + '\n';
  };

  return buildShape(character);
}

File: solutions/ruby/diamond/1/diamond.rb
class Diamond
    def self.make_diamond(character)
      width = (character.ord - 'A'.ord) * 2 + 1
  
      front = ('A'..character).map { |letter| build_line(letter, width) }
  
      (front + front[0...-1].reverse).join("\n") + "\n"
    end
  
    def self.build_line(letter, width)
      inner_width = (letter.ord - 'A'.ord) * 2 - 1
      outer_width = (width - (letter.ord - 'A'.ord) * 2 - 1) / 2
      letter_str = letter
      outer_pad = ' ' * outer_width
  
      content = inner_width < 1 ? letter_str : "#{letter_str}#{' ' * inner_width}#{letter_str}"
  
      "#{outer_pad}#{content}#{outer_pad}"
    end
end